docs/DeveloperGuide.adoc

```diff
@@ -142,105 +142,706 @@ image::StorageClassDiagram.png[]
+undo and redo capabilities by extending the `HistoryModel` which has the following operations:
 
-Step 3. The user executes `add n/David ...` to add a new person. The `add` command also calls `Model#commitAddressBook()`, causing another modified address book state to be saved into the `addressBookStateList`.
+* `Model#getHistoryManager()` -- Gets the `HistoryManager` instance.
+* `Model#setHistoryManager(HistoryManager)` -- Resets the `HistoryManager` data to the given `HistoryManager` in the argument.
+* `Model#getAvailableNumberOfExecutedCommands()` -- Gets the maximum available number of commands that can be undone.
+* `Model#getAvailableNumberOfInverselyExecutedCommands()` -- Gets the maximum available number of commands that can be redone.
+* `Model#canRollback()` -- Checks if it is possible to undo a command at the given state.
+* `Model#canCommit()` -- Checks if it is possible to redo a command at the given state.
+* `Model#rememberExecutedCommand(Command)` -- Remembers the given `Command` and stores it in `executedCommands` to facilitate undo capability for this command.
+* `Model#rememberInverselyExecutedCommand(Command)` -- Remembers the given `Command` and stores it in `inverselyExecutedCommands` to facilitate redo capability for this command.
+* `Model#rollback()` -- Inversely executes the latest command stored in `executedCommands` to revert the changes of the latest executed command made onto `Model`.
+* `Model#commit()` -- Executes the latest undone command stored in `inverselyExecutedCommands` to reapply the changes that were made onto `Model` by the latest undone command.
 
-image::UndoRedoState2.png[]
+Commands support the given operations to mutate the state of the `Model` and to check if they should be stored for undo/redo function:
 
-[NOTE]
-If a command fails its execution, it will not call `Model#commitAddressBook()`, so the address book state will not be saved into the `addressBookStateList`.
+* `Command#hasInverseExecution()` -- Checks if the command's execution mutates the state of the `Model`, which is used to determine if the command should be remembered by `HistoryManager`.
+* `Command#execute(Model)` -- Executes the command on the given `Model`.
+* `Command#executeInverse(Model)` -- Executes on the given `Model` such that it will undo whatever changes were made when `Command#execute(Model)` was called.
 
-Step 4. The user now decides that adding the person was a mistake, and decides to undo that action by executing the `undo` command. The `undo` command will call `Model#undoAddressBook()`, which will shift the `currentStatePointer` once to the left, pointing it to the previous address book state, and restores the address book to that state.
+Below is a class diagram between `Model`, `ModelManager`, `HistoryManager`, `CommandDeque` and `Command`.
 
-image::UndoRedoState3.png[]
+.Class Diagram for `Model`, `ModelManager`, `HistoryManager`, `CommandDeque` and `Command`
+image::HistoryManagerClassDiagram.png[]
```

Comment:
We don't think that there's a need to include all the methods of HistoryManager, CommandDeque etc, especially the getters.